File: src/main/java/udemy/stateful/Flink_Ex5_6_7_8_StreamOfTuples.java
package udemy.stateful;

import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.api.java.utils.ParameterTool;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;
import udemy.transformations.Average;
import udemy.transformations.ParseRow;
import udemy.transformations.SumAndCounts;
import udemy.utils.StreamUtil;

public class Flink_Ex5_6_7_8_StreamOfTuples {

    public static void main( String[] args ) throws Exception {
        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        final ParameterTool params = ParameterTool.fromArgs(args);
        env.getConfig().setGlobalJobParameters(params);

        DataStream<String> dataStream = StreamUtil.getDataStream(env, params);

        /*
        DataStream<Tuple2<String, Integer>> outStream1 = dataStream
                .map(new ExtractSpecialities())
                .flatMap(new SplitSpecial())
                .keyBy(tuple -> tuple.f0) // Stateful operation on each key
                .sum(1);

        DataStream<Tuple2<String, Double>> outStream2 = dataStream
                .map(new RowSplitter())
                .keyBy(tuple -> tuple.f0)
                .min(1);
         */

        DataStream<Tuple2<String, Double>> outStream3 = dataStream
                .map(new ParseRow.ParseRow6())
                .keyBy(tuple -> tuple.f0)
                .reduce(new SumAndCounts.SumAndCount1())
                .map(new Average());

        outStream3.print();

        env.execute("Tuple stream");
    }

    public static class SplitSpecial implements FlatMapFunction<String, Tuple2<String, Integer>> {
        public void flatMap(String input, Collector<Tuple2<String, Integer>> out) throws Exception {
            try {
                String[] specialities = input.split("\t");
                for (String speciality : specialities) {
                    out.collect(new Tuple2<>(speciality.trim(), 1));
                }
            } catch (Exception ignored) {
                System.out.println("Exception in SplitSpecial");
            }
        }
    }

    public static class RowSplitter implements MapFunction<String, Tuple2<String, Double>> {
        public Tuple2<String, Double> map(String row) throws Exception {
            try {
                String[] fields = row.split(" ");
                if(fields.length == 2) {
                    // Name, running time in minutes
                    return new Tuple2<>(fields[0], Double.parseDouble(fields[1]));
                }
            } catch (Exception ignored) {
                System.out.println("Exception in RowSplitter");
            }
            return null;
        }
    }

}

File: src/main/scala/stateful/Flink_Ex5_6_7_8_StreamOfTuples.scala
package stateful

import org.apache.flink.api.common.functions.{FlatMapFunction, ReduceFunction}
import org.apache.flink.api.java.utils.ParameterTool
import org.apache.flink.api.scala.createTypeInformation
import org.apache.flink.streaming.api.scala.{DataStream, StreamExecutionEnvironment}
import org.apache.flink.util.Collector
import utils.{EnvUtil, StreamUtil}

object Flink_Ex5_6_7_8_StreamOfTuples extends App {

  val (env: StreamExecutionEnvironment, params: ParameterTool) = EnvUtil.getEnv(args)

  val inputStream: DataStream[String] = StreamUtil.getDataStream(env, params)

  /*
  val outputStream1: DataStream[(String, Integer)] = extractSpecialities(inputStream)
    // flatMap needs to take the detour over an object
    .flatMap(FlattenSpecialities)
    .keyBy(_._1) // Stateful operation on each key
    .sum(1);

  val outputStream2: DataStream[String] = inputStream.map {
    line: String =>
      line.split(",")
        .map(_ (0))
        .mkString("")
  }

  val outputStream3: DataStream[(String, Double)] = inputStream.map {
    line: String =>
      val splitLine: Array[String] = line.split(" ")
      (splitLine(0), splitLine(1).toDouble)
  }
    .keyBy(_._1)
    .min(1);
   */

  // Course, length, area. Objective: Average course length in each area
  val outputStream4: DataStream[(String, Double, Int)] = inputStream.map {
    line: String =>
      val splitLine: Array[String] = line.split(",")
      (splitLine(2).trim, splitLine(1).toDouble, 1)
  }
    .keyBy(_._1)
  ////.reduce(SumAndCount) //// giving up with reduce, due to few docs around


  // Write to flink log
  outputStream4.print

  env.execute("Map FlatMap")
  
  object FlattenSpecialities extends FlatMapFunction[String, (String, Integer)] {
    override def flatMap(input: String, out: Collector[(String, Integer)]): Unit = {
      input.split("\t") foreach {
        speciality: String =>
          out.collect((speciality.trim, 1))
      }
    }
  }

  object SumAndCount extends ReduceFunction[(String, Double, Integer)] {
    override def reduce(
                         cumulative: (String, Double, Integer),
                         input: (String, Double, Integer)
                       ): (String, Double, Integer) = {
      (input._1, cumulative._2 + input._2, cumulative._3 + 1)
    }
  }

}
